signer-cloud-server/src/main/java/com/wultra/signercloud/server/document/DocumentService.java

```diff
@@ -188,17 +203,164 @@ private static byte[] getFileBytes(final MultipartFile file) {
+                .orElseThrow(() -> new DocumentNotFoundException("Document not found for document ID: " + documentId));
+
+        final var documentContent = documentContentRepository.findById(document.getDocumentContentId())
+                .orElseThrow(() -> new DocumentNotFoundException("Document content not found for document ID: " + documentId));
+
+        final var signer = signerRepository.findById(document.getSignerId())
+                .orElseThrow(() -> new SignerNotFoundException("Signer not found for document ID: " + documentId));
+
+        verifyDocumentCanBeSigned(signer, document);
+
+        final var signature = requestBody.signature();
+        final var signedDocumentBytes = verifySignatureAndSignDocument(signer.getCertificate(),
+                document.getHash(),
+                signature,
+                documentContent.getContent(),
+                configurationProperties.getSignatureHashAlgorithm());
+
+        final var updatedDocumentContent = documentContent.toBuilder()
+                .content(signedDocumentBytes)
+                .build();
+
+        documentContentRepository.save(updatedDocumentContent);
+
+        final var updatedDocument = document.toBuilder()
+                .timestampLastUpdated(Instant.now())
+                .fileSize(signedDocumentBytes.length)
+                .status(DocumentStatus.SIGNED)
+                .signature(signature)
+                .build();
+
+        documentRepository.save(updatedDocument);
+
+        final var downloadUrl = buildDocumentDownloadUri(documentId);
+        return new SignDocumentResponse(documentId, downloadUrl);
+    }
+
+    private void verifyDocumentCanBeSigned(final Signer signer, final Document document) {
+        if (SignerStatus.ACTIVE != signer.getStatus()) {
+            throw new SignDocumentException("Signer is not active. Signer: " + signer.getExternalSignerId());
+        }
+
+        if (DocumentStatus.WAITING != document.getStatus()) {
+            throw new SignDocumentException("Document is not in state when it can be signed");
+        }
+
+        final var waitingTimeout = configurationProperties.getWaiting().getTimeout();
+        if (waitingTimeout != null) {
+            final var documentSigningDeadline = document.getTimestampCreated().plus(waitingTimeout);
+            if (Instant.now().isAfter(documentSigningDeadline)) {
+                throw new SignDocumentException("Document signing timeout exceeded");
+            }
+        }
+    }
+
+    private byte[] verifySignatureAndSignDocument(final String certificateBase64,
+                                                         final String hashBase64,
+                                                         final String hashSignatureBase64,
+                                                         final byte[] documentBytes,
+                                                         final DigestAlgorithm signatureAlgorithm) throws CertificateException {
```

Comment:
We do not have conventions for that ye,t and it is difficult to edit IDE formatter, but I would write:

```suggestion
    private byte[] verifySignatureAndSignDocument(
            final String certificateBase64,
            final String hashBase64,
            final String hashSignatureBase64,
            final byte[] documentBytes,
            final DigestAlgorithm signatureAlgorithm) throws CertificateException {
```

Reply:
Yep, it looks better. Updated.